ocaml/forkexecd/lib/fe_stubs.c

```diff
@@ -0,0 +1,899 @@
+    for (int i = 0; i < 3; ++i) {
+        int fd = -1;
+        if (standard_fd_values[i] != Val_none) {
+            fd = Int_val(Some_val(standard_fd_values[i]));
+            if (fd < 0) {
+                log_fail("invalid standard fd");
+                mapped_logs_close(logs);
+                unix_error(EINVAL, "safe_exec", Nothing);
+            }
+        }
+        standard_fds[i] = fd;
+    }
+
+    FOREACH_LIST(arg, args) {
+        strings_size += strlen(String_val(Field(arg, 0))) + 1;
+        ++num_args;
+    }
+    if (num_args < 1) {
+        log_fail("no args");
+        mapped_logs_close(logs);
+        unix_error(EINVAL, "safe_exec", Nothing);
+    }
+
+    for (mlsize_t n = 0; n < num_environment; ++n)
+        strings_size += strlen(String_val(Field(environment, n))) + 1;
+
+    FOREACH_LIST(map, id_mapping) {
+        ++num_mappings;
+        value item = Field(map, 0);
+        const char *const uuid = String_val(Field(item, 0));
+        const int current_fd = Int_val(Field(item, 1));
+        // Check values.
+        // The check for UUID length makes sure we'll have space to replace last part of
+        // the command line arguments with file descriptor.
+        if (strlen(uuid) != 36 || current_fd < 0) {
+            log_fail("invalid mapping");
+            mapped_logs_close(logs);
+            unix_error(EINVAL, "safe_exec", Nothing);
+        }
+        strings_size += 36 + 1;
+    }
+    // Limit number of mappings to some sensible value.
+    // Also to avoid overflows with redirect code.
+    if (num_mappings > MAX_TOTAL_MAPPINGS) {
+        log_fail("too many mappings");
+        mapped_logs_close(logs);
+        unix_error(EINVAL, "safe_exec", Nothing);
+    }
+
+    // Prefix arguments with our helper to support syslog redirection.
+    if (key) {
+        num_args += 3;
+        strings_size += 64;
+        strings_size += strlen(key) + 1;
+    }
+
+    if (getrlimit(RLIMIT_NOFILE, &nofile_limit) < 0) {
+        log_fail("getrlimit error");
+        mapped_logs_close(logs);
+        unix_error(errno, "safe_exec", Nothing);
+    }
+
+    // Allocate buffer to copy all data that we need.
+    num_mappings += 3;
+    size_t total_size =
+        sizeof(exec_info) +
+        sizeof(char*) * (num_args + 1 + num_environment + 1) +
+        sizeof(mapping) * num_mappings +
+        strings_size;
+    exec_info *info = (exec_info *) caml_stat_alloc(total_size);
+    memset(info, 0, total_size);
+
+    log("copying");
+
+    // Copy all needed information.
+    {
+        char **ptrs = (char**)(info + 1);
+        mapping *mappings = (mapping *) (ptrs + num_args + 1 + num_environment + 1);
+        char *strings = (char*) (mappings + num_mappings);
+#ifdef FORKEXECD_DEBUG_LOGS
+        mapping *const initial_mappings = mappings;
+        char *const initial_strings = strings;
+#endif
+
+        // Copy arguments into list.
+        info->args = ptrs;
+        if (key) {
```

Comment:
This is potentially a stale pointer, there is a call to caml_stat_alloc above, which is part of the OCaml runtime, but 'key' points to memory inside an OCaml value, which could be moved by the garbage collector.

Reply:
Yes, attention was taken to not cause the garbage collector to trigger